src/algorithms/fardetectors/FarDetectorLinearTracking.cc

```diff
@@ -74,42 +80,65 @@ void FarDetectorLinearTracking::process(const FarDetectorLinearTracking::Input&
       trace("No hits in layer");
       return;
     }
-    convertedHits.push_back(ConvertClusters(*layerHits));
+    ConvertClusters(*layerHits, *assocHits, convertedHits, assocParts);
   }
 
   // Create a matrix to store the hit positions
   Eigen::MatrixXd hitMatrix(3, m_cfg.n_layer);
-  // Loop over all combinations of hits fitting a track to all layers
-  buildMatrixRecursive(m_cfg.n_layer - 1, &hitMatrix, convertedHits, outputTracks);
-}
 
-void FarDetectorLinearTracking::buildMatrixRecursive(
-    int level, Eigen::MatrixXd* hitMatrix, const std::vector<std::vector<Eigen::Vector3d>>& hits,
-    gsl::not_null<edm4eic::TrackSegmentCollection*> outputTracks) const {
+  // Create vector to store indexes of hits in the track
+  std::vector<int> layerHitIndex(m_cfg.n_layer, 0);
 
-  // Iterate over hits in this layer
-  for (auto hit : hits[level]) {
+  int layer = 0;
 
-    hitMatrix->col(level) << hit;
+  // Iterate over all combinations of measurements in the layers without recursion
+  while (layer >= 0) {
+    hitMatrix.col(layer) << convertedHits[layer][layerHitIndex[layer]];
```

Comment:
So the reason why we have to recompute SVD in `checkHitCombination` is because `hitMatrix` is appended here. But it never is reset? Does this mean that algorithm depends on how hits are sorted in the layers?

Reply:
Maybe I'm confused, but shouldn't hitMatrix.col(layer) add an extra row (of verctors) every time? It's not an update, is it?

Reply:
`hitMatrix.col(layer)` selects the column of the Eigen matrix relevant to the layer of the detector, then `<<` sets the values of that column to the values in the appropriate `convertedHits` vector.

The Eigen use of `<<` doesn't feel entirely consistent with the rest of C++.

Reply:
Thanks for explaining. I believe, one can push (not replace) individual values with `operator<<` in Eigen, that's why I asked.